9.0 port l10n_ch_sepa to v9

alexis-via opened the pull request:
This is in fact a full re-write of the module, which is now based on account_banking_sepa_credit_transfer.

alexis-via committed:
Re-write l10n_ch_sepa for Odoo 9, building above account_banking_sepa_credit_transfer

alexis-via commented:
This PR depends on this other PR: https://github.com/OCA/bank-payment/pull/270

alexis-via commented:
FYI, I'm evaluating the opportunity to modify this PR to support both sepa credit transfer and sepa direct debit for switzerland, because it should not be difficult to support both (probably with the same module that would only depend on "account_banking_pain_base"... stay tuned.

fclementic2c commented:
test done with Odoo 9.0e :
missing access rights :  
- no button add to payment order on supplier invoice
- no Payment menus visible
  -> tests done with user admin

I managed to generate the attached xml file and test it with the test platform
[sct_PAY0001 (3).xml.zip](https://github.com/OCA/l10n-switzerland/files/297363/sct_PAY0001.3.xml.zip)

But I got 7 errors see below : 
[sct_PAY0001 (3).xml.txt.zip](https://github.com/OCA/l10n-switzerland/files/297367/sct_PAY0001.3.xml.txt.zip)

Please also find my test DB 
[test_ch_sepa_2016-06-03_10-03-43.zip](https://github.com/OCA/l10n-switzerland/files/297370/test_ch_sepa_2016-06-03_10-03-43.zip)

What do you think ?

alexis-via commented:
@fclementic2c In your test, you paid a BVR invoice and a non-BVR invoice... I don't know how we are supposed to handle this particular case.

fclementic2c commented:
This is a very usual use case that we have to handle.

give us a call please (after lunch)

alexis-via committed:
Add support for direct debit pain.008.001.02.ch.01

alexis-via committed:
Proper handling of local_instrument: makes bvr a per-line setting and not a global setting

alexis-via committed:
Should now generate VALID pain.001.001.03.ch.02 files

alexis-via committed:
Follow a method rename in OCA/bank-payment

fclementic2c commented:
functional tests :+1: 

But what about partial payment and multi-currency ? Do you want to fix this in the other PR ?

alexis-via committed:
Remove useless inherit of invoice form view (which broke the inherit in the module account_payment_partner)

alexis-via committed:
Remove a method that has nothing to do in a localization module, and conflicts with account_payment_partner

yvaucher commented on l10n_ch_sepa/__openerp__.py:
```diff
@@ -1,62 +1,26 @@
     "test": [
-        "test/pain001_eu.yml",
-        "test/pain001_ch.yml",
+        # "test/pain001_eu.yml",
```
pain001_eu.yml can be completely removed

yvaucher commented on l10n_ch_sepa/models/account_payment_order.py:
```diff
@@ -0,0 +1,108 @@
+            for line in self.payment_line_ids:
+                if line.communication_type == 'bvr':
+                    sepa = False
+                    break
```
What about

```
      sepa &= not any(line.communication_type == 'bvr'
                      for line in self.payment_line_ids)
```

yvaucher commented on l10n_ch_sepa/models/account_payment_order.py:
```diff
@@ -0,0 +1,108 @@
+            if not partner_bank.ccp:
+                raise UserError(_(
+                    "The field 'CCP/CP-Konto' is not set on the bank "
+                    "'%s'.") % partner_bank.bank_id.ccp)
```
Here you probably want to display the bank name instead of the ccp which will be null.

yvaucher commented on l10n_ch_sepa/views/account_payment_line.xml:
```diff
@@ -0,0 +1,21 @@
+  License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+-->
+<odoo>
+<data>
```
`<data>` can me remove

yvaucher commented:
Nicely done

Few remarks of style taste but this is :+1: 

Travis error is due to unmerged dependancy in bank-payment https://github.com/OCA/bank-payment/pull/270

yvaucher commented on l10n_ch_base_bank/views/invoice.xml:
```diff
@@ -9,11 +9,24 @@
         <field name="date_due" position="after">
-          <field name="reference_type" nolabel="1"  attrs="{'readonly':[('state','!=','draft')]}"/>
-          <field name="reference" nolabel="1"  attrs="{'readonly':[('state','!=','draft')]}"/>
+          <field name="reference_type" nolabel="1"/>
```
Why was it removed ? Was it for test purpose ? I think we want to keep the readonly state once the invoice has been validated.

alexis-via replied:
In the field definition in the official account module, those fields are already readonly when not in draft, cf 
https://github.com/odoo/odoo/blob/9.0/addons/account/models/account_invoice.py#L213

alexis-via committed:
Take into account the remarks of Yannick Vaucher :

Fix error message
Use any() instead of a loop
remove <data> in view
Remove old test

alexis-via committed:
Add a know issue in readme about the DD file

alexis-via committed:
Add unittest

yvaucher commented:
Some lint in unittests:

```
l10n_ch_sepa/tests/test_ch_sct.py:64:80: E501 line too long (80 > 79 characters)
l10n_ch_sepa/tests/test_ch_sct.py:68:9: F841 local variable 'eur_currency_id' is assigned to but never used
l10n_ch_sepa/tests/test_ch_sct.py:140:80: E501 line too long (81 > 79 characters)
```

alexis-via committed:
PEP8 fixes

alexis-via commented:
@yvaucher remarks made me see that I have another issue to fix: the tests depend on account_banking_sepa_credit_transfer whereas the module only depend on account_banking_pain_base... I need to think about it...

alexis-via committed:
Migrate to a new architecture with 3 modules for PAIN

alexis-via committed:
Use setUp method in unit-test

alexis-via committed:
Fix BIC lenght

alexis-via committed:
Add a unit-test for payment type 3

Fix sepa compute method

alexis-via committed:
Fix BIC length

alexis-via committed:
Fix unit-test suite of l10n_ch_base_bank

alexis-via commented:
Travis will become green on this PR when this other PR is merged: https://github.com/OCA/bank-payment/pull/278

fclementic2c commented:
perfect thanks
:+1:

yvaucher commented:
Restarted travis jobs

alexis-via commented:
The PR in bank-payment has been merged ; Travis is green now !

yvaucher commented:
@alexis-via Thanks for this refactoring of the sepa module !
